vim-engine/src/main/kotlin/com/maddyhome/idea/vim/api/VimOptionGroupBase.kt

```diff
@@ -190,18 +221,548 @@ public abstract class VimOptionGroupBase : VimOptionGroup {
+ * For example, the `'wrap'` Vim option is not implemented by IdeaVim at all. Soft wraps must be implemented by the host
+ * editor. Similarly, IntelliJ has options that correspond to `'scrolloff'`, although the implementation is different to
+ * Vim's. IdeaVim might be able to use the same values while providing a different implementation.
+ *
+ * Unless the Vim value is explicitly set, the IDE value should take precedence. This allows users to opt in to Vim
+ * behaviour (`:set` or `~/.ideavimrc`), while still using the IDE to change settings. If the option has not been
+ * explicitly set, then it has a default Vim value, but the effective value comes from the IDE. When set via the Vim
+ * `:set` commands, the IDE value is updated to match. The user is free to update the value in the IDE and this is still
+ * reflected in the Vim option value, but treated internally as an external changed.
+ *
+ * When setting the effective value of local options, the global value is also updated. If a user opts in to modifying a
+ * Vim option, the global value is also considered explicitly set and this is copied to any new windows during
+ * initialisation, meaning new windows match the behaviour of the current window.
+ *
+ * Note that this class is an implementation detail of [VimOptionGroupBase] and derived instances, but cannot be
+ * made into a protected nested class because it is used by private helper classes.
+ */
+public sealed class OptionValue<T : VimDataType>(public open val value: T) {
+  /**
+   * The option value has been set as a default value by IdeaVim
+   *
+   * When setting an option, the value is a Vim default value. When getting a default option, the value might come from
+   * an IDE setting, but still uses the [Default] wrapper type.
+   */
+  public class Default<T : VimDataType>(override val value: T): OptionValue<T>(value)
+
+  /**
+   * The option value has been explicitly set by the user, by Vim commands
+   *
+   * The value has been set using the `:set` commands. When getting a value, this type is used if the value has been
+   * explicitly set by the user and the corresponding IDE setting (if any) still has the same value.
+   */
+  public class User<T : VimDataType>(override val value: T): OptionValue<T>(value)
+
+  /**
+   * The option value has been explicitly set by the user, but changed through the IDE
+   *
+   * This type is only used if the option has previously been set by the user using Vim's `:set` commands, but the
+   * current corresponding IDE setting no longer has the same value. This means that the user has explicitly set the
+   * option via Vim, but changed it in the IDE. If this value is used to initialise an option in a new window, it is
+   * treated as though the user explicitly set the option using Vim's `:set` commands.
+   */
+  public class External<T : VimDataType>(override val value: T): OptionValue<T>(value)
```

Comment:
As with the other comment, if I have `set nu` and I turn off the setting in IJ, I don't observe that the IJ option somehow affects numbering, not in the opened editors, nor in the newly created.

Reply:
If you have the Vim option set, you need to modify the local IntellIJ setting with _View | Active Editor_. Changing the global IntelliJ value in the settings dialog has no effect because the local value set by Vim overrides it.

This text makes an assumption about how the change is made in the IDE. I'll update the text.

Reply:
Updated comments to be more explicit that this refers to the IDE setting's local values, and how this works with the IDE's global value.